src/test/coll.rs

```diff
@@ -545,3 +543,120 @@ async fn ns_not_found_suppression() {
     coll.drop(None).await.expect("drop should not fail");
     coll.drop(None).await.expect("drop should not fail");
 }
+
+async fn delete_hint_test(options: Option<DeleteOptions>, name: &str) {
+    let _guard = LOCK.run_concurrently().await;
+
+    let client = EventClient::new().await;
+    let coll = client.database(name).collection(name);
+    let _: Result<DeleteResult> = coll.delete_many(doc! {}, options.clone()).await;
+
+    let events = client.get_command_started_events("delete");
+    assert_eq!(events.len(), 1);
+
+    let event_hint = events[0].command.get("hint").cloned();
+    let expected_hint = match options {
+        Some(options) => options.hint.map(|hint| hint.to_bson()),
+        None => None,
+    };
+    assert_eq!(event_hint, expected_hint);
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn delete_hint_keys_specified() {
+    let options = DeleteOptions::builder().hint(Hint::Keys(doc! {})).build();
+    delete_hint_test(Some(options), function_name!()).await;
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn delete_hint_string_specified() {
+    let options = DeleteOptions::builder()
+        .hint(Hint::Name(String::new()))
+        .build();
+    delete_hint_test(Some(options), function_name!()).await;
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn delete_hint_not_specified() {
+    delete_hint_test(None, function_name!()).await;
+}
+
+async fn find_one_and_delete_hint_test(options: Option<FindOneAndDeleteOptions>, name: &str) {
+    let _guard = LOCK.run_concurrently().await;
+    let client = EventClient::new().await;
+
+    let req = VersionReq::parse("< 4.2").unwrap();
```

Comment:
I think this ought to be `"<=4.2"`, since this should fail in 4.2 as well. Kind of curious why this isn't causing errors right now.

Reply:
I have `< 4.2` because the spec change you linked indicates that the driver should only raise an error (i.e. not send the command to the server) for servers < 4.2 -- "Older servers >= 4.2 will report an error for using this option." This test is only checking that the server received the command with the option specified, which is correct behavior for 4.2 even though the server will return an error. The test below (`find_one_and_delete_hint_server_version`) ensures that `find_one_and_delete` returns the correct thing for different server versions (i.e. driver error for < 4.2, server error for == 4.2, ok for 4.4+).

Reply:
Ah ok, makes sense :+1: